javascript/ql/src/semmle/javascript/security/performance/ReDoSUtil.qll

```diff
@@ -871,21 +872,31 @@ private module SuffixConstruction {
    * `prev` to `next` that the character symbol `char`.
    */
   predicate deltaClosedChar(StateInPumpableRegexp prev, string char, StateInPumpableRegexp next) {
-    char = relevant() and
-    deltaClosed(prev, getAnInputSymbolMatching(char), next)
+    deltaClosed(prev, getAnInputSymbolMatchingRelevant(char), next)
+  }
+
+  pragma[noinline]
+  InputSymbol getAnInputSymbolMatchingRelevant(string char) {
+    char = relevant(_) and
+    result = getAnInputSymbolMatching(char)
   }
 
   /**
-   * Gets a char used for finding possible suffixes.
+   * Gets a char used for finding possible suffixes inside `root`.
    */
-  private string relevant() { result = CharacterClasses::getARelevantChar() }
+  pragma[noinline]
+  private string relevant(RegExpRoot root) {
+    result = ["a", "9", "|", "\n", " "]
```

Comment:
`["a", "9", "|", "\n", " "]`: where do these magic constants come from?

Reply:
(Answered in top-level comment further down).